### ISSUE TITLE
result_backend cannot be configured in pytest 

### ISSUE THREAD

**Issue by cadupont** [celery/celery#4032]
## Checklist

- [x] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).
- [x] I have verified that the issue exists against the `master` branch of Celery.

----------------------------------------------
celery -A firehose.tasks.orchestration report

software -> celery:4.0.2 (latentcall) kombu:4.0.2 py:3.5.2
            billiard:3.5.0.2 py-amqp:2.1.4
platform -> system:Darwin arch:64bit imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:pyamqp results:redis://localhost:6379/0

result_backend: 'redis://localhost:6379/0'
broker_url: 'amqp://guest:********@localhost:5672//'
----------------------------------------------

## Steps to reproduce
Create one celery task in a separate file, import the celery task and try to get its result from within a pytest test. Try specifying the result backend via marks or fixtures.
e.g.
```
from ... import fetch
import pytest

@pytest.mark.celery(result_backend='redis://localhost:6379/0')
def test_successful_orchestration(celery_worker): 
    f = fetch.delay()
    assert f.get() == 'result'
```
I also have the same result if I try with the config fixture
```
@pytest.fixture(scope='session')
def celery_config():
    return {
        "result_backend": "redis://localhost/0",
        "broker_url": "amqp://guest:guest@localhost:5672//"
    }
```

## Expected behavior
I would expect the result_backend to be available to the test.
Note that when I launch a worker separately, when I am not testing, with the same backend settings, it works.

## Actual behavior
Running pytest mytest.py returns:

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../.pyenv/versions/3.5.2/envs/celery3.5.2/lib/python3.5/site-packages/celery/result.py:194: in get
    on_message=on_message,
../../.pyenv/versions/3.5.2/envs/celery3.5.2/lib/python3.5/site-packages/celery/backends/base.py:466: in wait_for_pending
    no_ack=no_ack,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <celery.backends.base.DisabledBackend object at 0x1103d5518>, args = ('985a0f2e-ff3a-4ada-a045-11137f358694',), kwargs = {'interval': 0.5, 'no_ack': True, 'on_interval': <promise@0x1105f0048>, 'timeout': None}

    def _is_disabled(self, *args, **kwargs):
>       raise NotImplementedError(E_NO_BACKEND.strip())
E       NotImplementedError: No result backend is configured.
E       Please see the documentation for more information.

../../.pyenv/versions/3.5.2/envs/celery3.5.2/lib/python3.5/site-packages/celery/backends/base.py:772: NotImplementedError

**Comment by georgepsarakis** [comment]
Not sure if I fully understand the issue, but you can see how the broker and result backend is configured for running the tests in the [tox.ini](https://github.com/celery/celery/blob/master/tox.ini#L33) file.

**Comment by cadupont** [comment]
Here's what I found after a little bit of debugging.
[From the docs](http://docs.celeryproject.org/en/master/userguide/testing.html):
```
def test_add(celery_worker):
    mytask.delay()
```

I expect {celery_worker} will contain the relevant result {backend} object inside of its {app} object.

However, "mytask" does not relate to the same app as the worker's app, i.e. the result backend on becomes a DisabledBackend object (instead of a RedisBackend object on the worker's app). (the app is not the same as soon as you are in celery/local.py)

Am I missing something obvious? From the docs I would have expected to be able to call "mytask" (a task defined in a random and external module) and for it to be picked up by celery_worker using configs defined either via fixtures or marks.

**Comment by thedrow** [comment]
You should override the `celery_app` fixture to use your app.
See [here](https://github.com/celery/celery/blob/master/celery/contrib/pytest.py#L164).
Let us know if you still have problems and I'll reopen this issue.

**Comment by namezys** [comment]
I actually don't understand what you mean?

I'm trying to write something like this:
```
@pytest.fixture
def celery_app(celery_app):
    return celery_app


@pytest.fixture(scope='session')
def celery_config():
    return {
        'broker_url': 'redis://localhost:6379/0',
        'result_backend': 'rpc',
    }


def test_blablabla(celery_worker):
    print celery_worker
    r = tasks.task_for_test.delay()
    while not r.ready():
        pass
```

And got error:

```
self = <celery.backends.base.DisabledBackend object at 0x116354750>, task_id = 'b8a0e736-65d4-4324-b35c-d58da8bc1c40', cache = True

    def get_task_meta(self, task_id, cache=True):
        self._ensure_not_eager()
        if cache:
            try:
                return self._cache[task_id]
            except KeyError:
                pass

>       meta = self._get_task_meta_for(task_id)
E       AttributeError: 'DisabledBackend' object has no attribute '_get_task_meta_for'
```

Looks like I can't setup result backend.

**Comment by thedrow** [comment]
Oh you're trying to override celery config to tell it to use a results backend.
Right, The workaround is to currently return the App() instance you defined in your code in the celery_app fixture.

E.g.
```
from myapp.tasks import my_celery_app
import pytest

@pytest.fixture
def celery_app():
  return my_celery_app
```
and override the configuration on the celery app you defined.

**Comment by namezys** [comment]
Yep. After 2 hours with debugger and digging into the celery I understood it.
It's not that I expected to get. This looks useless.
It looks bad because I should override config for `my_celery_app`. But I don't see easy way to fix or workaround it.
I think it will be a good idea to append this case into documentary.

Thank you.

**Comment by thedrow** [comment]
We can add it to the documentation as a known issue until we fix it. PRs welcome :)
